portfolio/src/main/webapp/script.js

```diff
@@ -60,6 +51,61 @@ async function getCommentFromFetch() {
         var para = document.createElement("p");
         var node = document.createTextNode(commentString);
         para.appendChild(node);
+        var button = document.createElement("button");
+        button.innerHTML = "Like";
+        const input = commentString.slice();
+        button.addEventListener("click", function () { likeComments(input) });
         commentContainer.appendChild(para);
+        commentContainer.appendChild(button);
     };
+}
+async function getCommentFromFetch() {
+    console.log("Try fetch");
+    const response = await fetch("/data");
+    const comments = await response.json();
+    const dummyElement = comments.pop();
+    var maxComments = parseInt(dummyElement.timestamp, 10);
+    if (Number.isNaN(maxComments)) {
+        console.log("You are trying to input a non-numeric data, please try again");
+        // Set the limit to default value
+        maxComments = 10;
+    }
+    const commentLimitContainer = document.getElementById("comment-limit");
+    commentLimitContainer.value = (maxComments.toString());
+    fetchHelper(maxComments, comments);
+}
+
+async function onChangeFromFetch() {
+    console.log("Try fetch");
+    document.getElementById("comment.")
+    const response = await fetch("/data");
+    const comments = await response.json();
+    const dummyElement = comments.pop();
+    var maxComments = parseInt(document.getElementById("comment-limit").value, 10);
+    if (Number.isNaN(maxComments)) {
+        console.log("You are trying to input a non-numeric data, please try again");
+        // Set the limit to default value
+        maxComments = 10;
+    }
+    if (maxComments < 1) {
+        console.log("Please input a positive integer");
+        // Set the limit to default value
+        maxComments = 10;
+    }
+    fetchHelper(maxComments, comments);
+}
+
+async function deleteComments() {
+    console.log("Delete all comments in datastore");
+    const request = new Request("/delete-data", {method: "POST"});
+    const response = await fetch(request);
+    const commentContainer = document.getElementById("comment-container");
+    commentContainer.innerHTML = "";
+}
+
+async function likeComments(commentString) {
+    console.log("Like a certain comment");
+    const queryURL = "/like-data?comment=" + commentString;
```

Comment:
I would suggest referring to comments by the ID assigned to them in Datastore.

Consider a comment like "me&you=friends". With simple string concatenation here, the queryURL becomes "/like-data?comment=me&you=friends", causing the backend to search for a comment of "me" and a meaningless "you" parameter.

Reply:
At first I did consider using the ID that is assigned to the comment in Datastore. But in the js file I don't know how to get the ID of the comment. The only method that I came up with is to send a GET request to the servlet file and get the ID as the response.

Reply:
I think I just came up with a method to get the ID of the comment. Maybe it's a little bit hard but I will have a try.

Reply:
As discussed over chat, https://cloud.google.com/appengine/docs/standard/java/datastore/creating-entity-keys?hl=en details how to pass around string-versions of keys to do these lookups. And we probably don't need any new backend methods to provide the keys of the comments, but should prefer to modify the existing backend that returns comments.

Reply:
I've successfully passed the key to the servlet and updated other files to accomodate the change.